exec/java-exec/src/main/java/org/apache/drill/exec/server/FailureUtils.java

```diff
@@ -0,0 +1,97 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.server;
+
+import io.netty.util.internal.OutOfDirectMemoryError;
+import org.apache.drill.common.CatastrophicFailure;
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.exec.exception.OutOfMemoryException;
+
+public final class FailureUtils {
+  private static volatile boolean unrecoverableFailure;
+
+  public static final int EXIT_CODE_HEAP_OOM = -1;
+  public static final int EXIT_CODE_JAVA_ERROR = -2;
+
+  /**
+   * This message is used to distinguish between direct memory and heap memory {@link OutOfMemoryError}s.
+   */
+  public static final String DIRECT_MEMORY_OOM_MESSAGE = "Direct buffer memory";
+
+  private FailureUtils() {
+    // Don't instantiate
+  }
+
+  public static boolean isDirectMemoryOOM(Throwable e) {
+    if (e instanceof OutOfDirectMemoryError || e instanceof OutOfMemoryException) {
+      // These are always direct memory errors
+      return true;
+    }
+
+    return (e instanceof OutOfMemoryError) && DIRECT_MEMORY_OOM_MESSAGE.equals(e.getMessage());
+  }
+
+  public static boolean isHeapOOM(Throwable e) {
+    if (e instanceof OutOfMemoryError) {
+      return !isDirectMemoryOOM(e);
+    } else {
+      return false;
+    }
+  }
+
+  public static void unrecoverableFailure(Throwable e, String message, int exitCode) {
+    unrecoverableFailure = true;
+    CatastrophicFailure.exit(e, message, exitCode);
+  }
+
+  public static boolean hadUnrecoverableFailure() {
+    return unrecoverableFailure;
+  }
+
+  public static void setUncaughtExceptionHandler() {
+    Thread.currentThread().setUncaughtExceptionHandler(new FailureUtils.DrillExceptionHandler());
+  }
+
+  private static class DrillExceptionHandler implements Thread.UncaughtExceptionHandler {
+    private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(Drillbit.class);
+
+    @Override
+    public void uncaughtException(Thread t, Throwable e) {
+      handleException(e);
+    }
+
+    public static void handleException(Throwable e) {
+      if (isHeapOOM(e)) {
+        // We can't recover cleanly from these.
+        unrecoverableFailure(e, "Unable to handle heap OOM.", EXIT_CODE_HEAP_OOM);
+      } else if (isDirectMemoryOOM(e)) {
+        // It's a problem, but we might be able to recover when we free memory.
+        UserException.memoryError(e).build(logger);
+      } else if (e instanceof Error) {
```

Comment:
Note that `AssertionError` is an instance of an `Error` and it triggers `unrecoverableFailure()` that always bypass Drillbit close sequence. Is it intentional?

Reply:
@vrozov I didn't think of this. What do you think the right way to handle it is? In some cases I think we could still gracefully shutdown after an assert fails. But in other cases an AssertionError may occur because the Drillbit's state is corrupt and it cannot reliably perform a graceful shutdown.